pkg/ipcache/ipcache.go

```diff
@@ -473,16 +473,42 @@ func (ipc *IPCache) RemoveMetadata(prefix netip.Prefix, resource ipcacheTypes.Re
 // labels with these prefixes, thereby making these prefixes selectable in
 // policy via local ("CIDR") identities.
 //
+// It will return a list of identities corresponding the supplied prefixes.
+// The identity for a given prefix may be nil, as allocation may fail. Callers
+// should handle this accordingly.
+//
 // This will trigger asynchronous calculation of any datapath updates necessary
 // to implement the logic associated with the new CIDR labels.
-func (ipc *IPCache) UpsertPrefixes(prefixes []netip.Prefix, src source.Source, resource ipcacheTypes.ResourceID) {
+func (ipc *IPCache) UpsertPrefixes(prefixes []netip.Prefix, src source.Source, resource ipcacheTypes.ResourceID) []*identity.NumericIdentity {
```

Comment:
Agree we have an eventual consistency model, though with a bit less complexity in the ipcache interface I think?

Step (1) allocate CIDR identities only occurs async as part of step (4) from the `InjectLabels()` call in main today but yeah I agree it's a bit odd even today because the policy calculation is triggered separately from the identity allocation. So the endpoint regeneration -> policy calculation step may calculate policy without taking into account the new identities for the new prefixes, then those new identities/policy impacts will only be applied at step (5).